plugins/chromatic_correction.py

```diff
@@ -0,0 +1,346 @@
+# Based on diffraction, bead size in the image appears bigger which is used in calculation for finding and matching
+# bead centers. The actual bead diameter is multiplied by a scaling factor
+# The scaling is based on the provided dataset
+SCALE = 5.5
+# Maximum number of beads to find in the given image
+BEAD_QUANTITY = 15
+
+
+def find_corresponding_points(centers_1: numpy.ndarray, centers_2: numpy.ndarray, max_distance: float):
+    """
+    Find corresponding points in two sets of points that are within a specified maximum distance.
+    :param centers_1: Nx2 array of points (x, y) in the first image.
+    :param centers_2: Mx2 array of points (x, y) in the second image.
+    :param max_distance: (float in pixels) Maximum allowable distance for point pairs.
+    :returns:
+        pairs: (numpy.ndarray) Corresponding point pairs (index1, index2).
+    """
+    # Calculate pairwise distances
+    distances = cdist(centers_1, centers_2)
+
+    # Find pairs within the max_distance
+    all_pairs = numpy.argwhere(distances <= max_distance)
+    sorted_pairs = sorted(all_pairs, key=lambda x: distances[x[0], x[1]])
+
+    matched_1 = set()
+    matched_2 = set()
+    pairs = []
+
+    for i, j in sorted_pairs:
+        if i not in matched_1 and j not in matched_2:
+            pairs.append([i, j])
+            matched_1.add(i)
+            matched_2.add(j)
+
+    return numpy.array(pairs)
+
+
+def estimate_transformation_from_pointset(points1: numpy.ndarray, points2: numpy.ndarray):
+    """
+    Estimate the affine transformation values from a set of corresponding points that maps `points2` to `points1`.
+    :param points1: Nx2 (x, y) location of features
+    :param points2: Nx2 (x, y) location of the corresponding features
+    :returns:
+        tuple: A tuple containing correction parameters:
+            - scale_cor ((float, float)): Scaling factors along the x and y axes
+            - translation_cor ((float, float)): Translation components along the x and y axes
+            - rotation_cor (float): Rotation angle (in radians)
+            - shear_cor (float): Shear angle (in radians)
+    """
+    affine = AffineTransform.from_pointset(points1, points2)
+    mat = numpy.eye(3)
+    mat[:2, :2] = affine.matrix
+    mat[:2, 2] = affine.translation
+    ref_mat = numpy.eye(3)
+    mat = ref_mat @ mat
+
+    # Scale is the norm of the first two columns
+    sx = numpy.linalg.norm(mat[:2, 0])
+    sy = numpy.linalg.norm(mat[:2, 1])
+    scale = numpy.array([sx, sy])
+    scale_cor = (1 / scale[0], 1 / scale[1])
+
+    # compute translation
+    tx = mat[0, 2]
+    ty = mat[1, 2]
+    translation_cor = (tx, -ty)
+
+    # Compute rotation
+    rotation_cor = numpy.arctan2(mat[1, 0] / sx, mat[0, 0] / sx)
+
+    # compute shear
+    a, b = mat[0, 0], mat[0, 1]
+    c, d = mat[1, 0], mat[1, 1]
+    shear_cor = numpy.arctan2(a * c + b * d, scale_cor[0] * scale_cor[1])
+
+    return scale_cor, translation_cor, rotation_cor, shear_cor
```

Comment:
I don't understand why you can't just use `affine.translation, affine.scale, affine.rotation, affine.shear`. 
If this is really different, can you add a comment explaining why?

Reply:
This function does two things - calculate the affine transformation parameters and then does correction on those parameters such that Odemis can display it through the correction metadata. To see if the first method works, I checked how the Odemis display images. I did not dwell a lot on the mathematics. 

For the uncorrected affine parameters, these calculated values do not match with the attributes of affine transform. The closest values come from this function alt_transformation_matrix_to_implicit(affine.matrix, "RSU"), in which rotation and scale matches but not shear.  

As chromatic aberration is a setup to test the approach, in my opinion making a separate function will help in faster debugging and fixing. This way we can ensure that the already used functions in util/transform.py are separate and are not modified for this single purpose.

Reply:
I have explained the same in the comment

Reply:
It could be that `alt_transformation_matrix_to_implicit()` contains bugs. I don't think it was extensively validated, especially wrt the shear.
Please compare to `acq.stream._base.getPixelCoordinates()` (does the reverse), and `util.conversion.get_img_transformation_md()`. It might be possible to directly use this second function directly, by creating a 3x3 matrix from the AffineTransform.matrix + translation. (if that works, maybe you could check that your current code and get_img_transformation_md() always behave the same, in which case you could replace the content of get_img_transformation_md() with your current code, which looks very neat).